### PR TITLE
CASMINST-4462 CASMINST-4805: NCN image modification procedure

### PR DESCRIPTION
**Pull request by heemstra**
Create a procedure for using CFS to modify NCN boot artifacts prior
to booting them. This procedure mostly guides the reader through
existing documentation, with only a few new steps added that are
specific to interacting with NCN artifacts.

### CODE REVIEW

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+1. [Import External Image to IMS](../image_management/Import_External_Image_to_IMS.md)
+
+    This document will instruct the admin to set several environment variables, including the three set in
+    the previous step.
+
+1. [Create and Populate a VCS Configuration Repository](Create_and_Populate_a_VCS_Configuration_Repository.md)
+
+   **NOTE:** if the image modification is a kernel-level change, a new `initrd` can be created by invoking
+   the following script: `/srv/cray/scripts/common/create-ims-initrd.sh`. This script is embedded in the
+   NCN SquashFS. After the script completes, a new `initrd` will be available at `/boot/initrd`. CFS will
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:37]:
```suggestion
   NCN SquashFS. After the script completes, a new `initrd` file will be available in the `/boot/` directory. CFS will
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+   automatically make this `initrd` available at the end of the CFS session.
+
+   **WARNING:** if you do not run the above script, **DO NOT** download the initrd or kernel after the
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:40]:
```suggestion
   **WARNING:** If the above script is not run, **DO NOT** download the `initrd` or kernel after the
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   CFS session completes (see below). CFS will collect `/boot/initrd` and `/boot/vmlinuz` from the modified
+   SquashFS. If you do not run the `create-ims-initrd.sh` script, these artifacts will be incorrect and
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:42]:
```suggestion
   SquashFS. If the `create-ims-initrd.sh` script is not run, then these artifacts will be incorrect and
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   will not be able to boot an NCN.
+
+   If you're not modifying anything at the kernel-level there is no need to create a new `initrd`. In
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:45]:
```suggestion
   When not modifying anything at the kernel-level, there is no need to create a new `initrd`. In
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+   that case, simply use the existing `initrd`.
+
+1. [Create a CFS Configuration](Create_a_CFS_Configuration.md)
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:48]:
```suggestion
1. Create a CFS configuration.

    See [Create a CFS Configuration](Create_a_CFS_Configuration.md).
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+1. [Create an Image Customization CFS Session](Create_an_Image_Customization_CFS_Session.md)
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:50]:
```suggestion
1. Create an image customization CFS session.

    See [Create an Image Customization CFS Session](Create_an_Image_Customization_CFS_Session.md).
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+1. Download the Resultant NCN Artifacts
+
+    **NOTE:** `$IMS_RESULTANT_IMAGE_ID` is the IMS image ID returned in the output of the last command
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:54]:
```suggestion
    > **NOTE:** `$IMS_RESULTANT_IMAGE_ID` is the IMS image ID returned in the output of the last command
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    in the previous step:
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:55]:
```suggestion
    > in the previous step:
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+
+    ```console
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:57]:
```suggestion
    > ```console
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    ncn# cray cfs sessions describe example --format json | jq .status.artifacts
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:58]:
```suggestion
    > ncn# cray cfs sessions describe example --format json | jq .status.artifacts
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+
+    ```console
+    ncn# cray artifacts get boot-images $IMS_RESULTANT_IMAGE_ID/rootfs kubernetes-<version>-1.squashfs
+
+    ncn# cray artifacts get boot-images $IMS_RESULTANT_IMAGE_ID/initrd initrd.img-<version>-1.xz
+
+    ncn# cray artifacts get boot-images $IMS_RESULTANT_IMAGE_ID/kernel 5.3.18-150300.59.43-default-<version>-1.kernel
+    ```
+
+1. Upload NCN boot artifacts into S3
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:69]:
```suggestion
1. Upload NCN boot artifacts into S3.
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+
+    This steps assumes that the `docs-csm` RPM is installed.
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:71]:
```suggestion
    This steps requires that the `docs-csm` RPM is installed. See [Check for Latest Documentation](../../update_product_stream/index.md#documentation).
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+
+    ```console
+    ncn# /usr/share/doc/csm/scripts/ceph-upload-file-public-read.py --bucket-name ncn-images --key-name 'k8s/<new-version>/filesystem.squashfs' --file-name kubernetes-<version>-1.squashfs
+
+    ncn# /usr/share/doc/csm/scripts/ceph-upload-file-public-read.py --bucket-name ncn-images --key-name 'k8s/<new-version>/initrd' --file-name initrd.img-<version>-1.xz
+
+    ncn# /usr/share/doc/csm/scripts/ceph-upload-file-public-read.py --bucket-name ncn-images --key-name 'k8s/<new-version>/kernel' --file-name <version>-1.kernel
+    ```
+
+1. Update NCN Boot Parameters
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:81]:
```suggestion
1. Update NCN boot parameters.
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+
+    Get the existing `metal.server` setting for the xname of the node of interest:
+
+    ```console
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:85]:
```suggestion
        ```console
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+    ncn# XNAME=<your-xname>
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:86]:
```suggestion
        ncn# XNAME=<your-xname>
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    ncn# METAL_SERVER=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' \
+         | awk -F 'metal.server=' '{print $2}' \
+         | awk -F ' ' '{print $1}')
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:89]:
```suggestion
             | awk -F ' ' '{print $1}')
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+    ```
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:90]:
```suggestion
        ```
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+
+    Verify the variable was set correctly: `echo $METAL_SERVER`
+
+    Update the kernel, initrd, and metal server to point to the new artifacts. Note that multiple Xnames can
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:94]:
```suggestion
    1. Update the kernel, `initrd`, and metal server to point to the new artifacts.

        Note: Multiple xnames can be updated at the same time, if desired.
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+    be updated at the same time if desired.
+
+    ```console
+    ncn# S3_ARTIFACT_PATH=ncn-images/k8s/<new-version>
+    ncn# NEW_METAL_SERVER=http://rgw-vip.nmn/$S3_ARTIFACT_PATH
+
+    ncn# PARAMS=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' | \
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:101]:
```suggestion
            ncn# PARAMS=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' | \
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+         sed "/metal.server/ s|$METAL_SERVER|$NEW_METAL_SERVER|")
+    ```
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:103]:
```suggestion
            ```
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+
+    Verify the value of `$NEW_METAL_SERVER` was set correctly: `echo $PARAMS`
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:105]:
```suggestion
        1. Verify that the value of `$NEW_METAL_SERVER` was set correctly:

            ```console
            ncn# echo $PARAMS
            ```
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+
+    In the following invocation, `$XNAME` may be one more more Xname.
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:107]:
```suggestion
        1. Update BSS.
        
            In the following invocation, `$XNAME` may be one or more xnames.
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+
+    ```console
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:109]:
```suggestion
            ```console
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 24]
```diff
@@ -0,0 +1,116 @@
...
+    ncn# cray bss bootparameters update --hosts $XNAME   \
+         --kernel "s3://$S3_ARTIFACT_PATH/kernel" \
+         --initrd "s3://$S3_ARTIFACT_PATH/initrd" \
+         --params "$PARAMS"
+    ```
```
Comment (mharding-hpe) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:114]:
```suggestion
            ```
```

**Review by mharding-hpe — COMMENTED** [review]
Going to fix some linting issues, then re-review.

**Review by mharding-hpe — APPROVED** [review]

**Review by rustydb — APPROVED** [review]
/backport main

**Review by rustydb — CHANGES_REQUESTED** [review]
Requires wipe step.

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 25]
```diff
@@ -0,0 +1,144 @@
+# Pre-Boot Configuration of NCN Images
+
+**NOTE:** Some of the documentation linked from this page mentions use of the BOS service. The use of BOS
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:3]:
```suggestion
**NOTE:** Some of the documentation linked from this page mentions use of the Boot Orchestration Service (BOS). The use of BOS
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 26]
```diff
@@ -0,0 +1,144 @@
...
+
+    1. Verify that the variable was set correctly.
+
+        ```console
+        ncn# echo $METAL_SERVER
+        ````
+
+    1. Update the kernel, `initrd`, and `metal.server` to point to the new artifacts.
+
+        Note: Multiple xnames can be updated at the same time, if desired.
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:113]:
```suggestion
        **NOTE:** Multiple xnames can be updated at the same time, if desired.
```
This is the formatting in several instances above.

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+            ncn# NEW_METAL_SERVER=http://rgw-vip.nmn/$S3_ARTIFACT_PATH
+
+            ncn# PARAMS=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' | \
+                 sed "/metal.server/ s|$METAL_SERVER|$NEW_METAL_SERVER|")
+            ```
+
+        1. Verify that the value of `$NEW_METAL_SERVER` was set correctly.
+
+            ```console
+            ncn# echo $PARAMS
```
[the remark this answers is not in the document]
Comment (heemstra) [reply operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md]: Thanks - added some clarification here.

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md (hunk 26 above)
Comment (heemstra) [reply operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md]: It seems this comment was made on out-of-date source.  The verbiage around 'multiple xnames' was changed in several places.

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 28]
```diff
@@ -0,0 +1,131 @@
...
+    ncn# S3_ARTIFACT_PATH=ncn-images/k8s/<new-version>
+    ncn# NEW_METAL_SERVER=http://rgw-vip.nmn/$S3_ARTIFACT_PATH
+
+    ncn# PARAMS=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' | \
+         sed "/metal.server/ s|$METAL_SERVER|$NEW_METAL_SERVER|")
+    ```
+
+    Verify the value of `$NEW_METAL_SERVER` was set correctly within the boot parameters: `echo $PARAMS`
+
+    In the following invocation, `$XNAME` may be one more more Xname.
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:120]:
```suggestion
    In the following invocation, `$XNAME` may be more than one xname.
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 29]
```diff
@@ -0,0 +1,131 @@
...
+    ```console
+    ncn# XNAME=<your-xname>
+    ncn# METAL_SERVER=$(cray bss bootparameters list --hosts $XNAME --format json | jq '.[] |."params"' \
+         | awk -F 'metal.server=' '{print $2}' \
+         | awk -F ' ' '{print $1}')
+    ```
+
+    Verify the variable was set correctly: `echo $METAL_SERVER`
+
+    Update the kernel, initrd, and metal server to point to the new artifacts. Note that multiple Xnames can
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:107]:
```suggestion
    Update the kernel, initrd, and metal server to point to the new artifacts. Note that multiple xnames can
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 30]
```diff
@@ -0,0 +1,131 @@
...
+
+   **WARNING:** if you do not run the above script, **DO NOT** download the `initrd` or `kernel` after the
+   CFS session completes (see below). CFS will collect `/boot/initrd` and `/boot/vmlinuz` from the modified
+   SquashFS. If you do not run the `create-ims-initrd.sh` script, these artifacts will be incorrect and
+   will not be able to boot an NCN.
+
+   If you're not modifying anything at the kernel-level there is no need to create a new `initrd`. In
+   that case, simply use the existing `initrd`.
+
+   **NOTE:** if desired, the default SSH *host* keys can be removed with a CFS Ansible task:
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:48]:
```suggestion
   **NOTE:** If desired, the default SSH *host* keys can be removed with a CFS Ansible task:
```

File: operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md [hunk 31]
```diff
@@ -0,0 +1,131 @@
...
+    ncn# export IMS_ROOTFS_FILENAME=<version>-filesystem.squashfs
+
+    ncn# export IMS_KERNEL_FILENAME=<version>-kernel
+
+    ncn# export IMS_INITRD_FILENAME=<version>-initrd
+    ```
+
+1. [Import External Image to IMS](../image_management/Import_External_Image_to_IMS.md)
+
+    This document will instruct the admin to set several environment variables, including the three set in
```
Comment (nrockershousen) [inline operations/configuration_management/Pre_Boot_NCN_Image_Configuration.md:30]:
```suggestion
    This document will instruct the administrator to set several environment variables, including the three set in
```

**Review by nrockershousen — APPROVED** [review]

**Pull request merged**